### PR TITLE
feat: add `bigframes.bigquery.json_set`

### PR DESCRIPTION
**Pull request by chelsea-lin**
BEGIN_COMMIT_OVERRIDE
feat: add `bigframes.bigquery.json_set` (#782)
END_COMMIT_OVERRIDE

This change is creating a `bigframes.bigquery.json_set` API, which can produces a new JSON value within a Series by inserting or replacing values at specified paths.
- [X] Make sure to open an issue as a internal bug 346813016
- [X] Ensure the tests and linter pass
- [X] Code coverage does not decrease (if any source code was changed)
- [X] Appropriate docs were updated (if necessary)

Fixes internal issue 346813016 🦕

### CODE REVIEW

File: bigframes/bigquery/__init__.py [hunk 1]
```diff
@@ -154,6 +158,64 @@ def array_to_string(series: series.Series, delimiter: str) -> series.Series:
...
+    """
+    # SQLGlot parser does not support the "create_if_missing => true" syntax, so
+    # create_if_missing is not currently implemented.
+
+    # Currently limited to single JSON path/value pairs (binary operations only).
+    if len(json_path_value_pairs) != 1:
+        raise ValueError(
+            "Expected exactly one JSON path and value pair but found "
+            + f"{len(json_path_value_pairs)} pairs."
+        )
```
Comment (TrevorBergeron) [inline bigframes/bigquery/__init__.py:205]: Why not just loop through the pairs and repeatedly apply?

File: bigframes/operations/__init__.py [hunk 2]
```diff
@@ -707,6 +707,19 @@ def output_type(self, *input_types):
...
+@dataclasses.dataclass(frozen=True)
+class JSONSet(BinaryOp):
+    name: typing.ClassVar[str] = "json_set"
+    json_path: str
+
+    def output_type(self, *input_types):
+        input_type = input_types[0]
+        if not dtypes.is_json_like(input_type):
+            raise TypeError("Input type must be an JSON or JSON-formatted string type.")
+        return input_type
```
Comment (TrevorBergeron) [inline bigframes/operations/__init__.py:720]: Are there any restrictions on the right argument? Or does it support every single datatype?

File: bigframes/core/compile/scalar_op_compiler.py [hunk 3]
```diff
@@ -894,6 +894,16 @@ def array_to_string_op_impl(x: ibis_types.Value, op: ops.ArrayToStringOp):
...
 
 
+# JSON Ops
+@scalar_op_compiler.register_binary_op(ops.JSONSet, pass_op=True)
+def json_set_op_impl(x: ibis_types.Value, y: ibis_types.Value, op: ops.JSONSet):
+    return vendored_ibis_ops.JSONSet(
+        x,
+        json_value=y,
+        json_path=op.json_path,
+    ).to_expr()
```
Comment (TrevorBergeron) [inline bigframes/core/compile/scalar_op_compiler.py:904]: is it possible to implement through the ibis udf interface, or is the syntax too non-standard? The new custom op will add extra work when upgrading to ibis 9

Comment (tswast) [reply bigframes/core/compile/scalar_op_compiler.py:904]: See: https://github.com/googleapis/python-bigquery-dataframes/blob/ec5b06856762e67060642cc106f43f0d202a08bf/bigframes/core/compile/scalar_op_compiler.py#L1462 for an example.

File: bigframes/bigquery/__init__.py (hunk 1 above)
Comment (chelsea-lin) [reply bigframes/bigquery/__init__.py]: Good points. It should be solved!

File: bigframes/operations/__init__.py (hunk 2 above)
Comment (chelsea-lin) [reply bigframes/operations/__init__.py]: Good points. Added additional type checking here.

File: bigframes/core/compile/scalar_op_compiler.py (hunk 3 above)
Comment (chelsea-lin) [reply bigframes/core/compile/scalar_op_compiler.py]: yes, it can be implemented through ibis udf interface.

**Review by TrevorBergeron — APPROVED** [review]
LGTM

**Pull request merged**